lib/elixir/lib/module/types/expr.ex

```diff
@@ -1,7 +1,477 @@
 defmodule Module.Types.Expr do
   @moduledoc false
 
-  def of_expr(_expr, _stack, context) do
+  import Module.Types.{Helpers, Infer}
+  import Module.Types.Pattern, only: [of_guard: 3, of_pattern: 3]
+
+  # :atom
+  def of_expr(atom, _stack, context) when is_atom(atom) do
+    {:ok, {:atom, atom}, context}
+  end
+
+  # 12
+  def of_expr(literal, _stack, context) when is_integer(literal) do
+    {:ok, :integer, context}
+  end
+
+  # 1.2
+  def of_expr(literal, _stack, context) when is_float(literal) do
+    {:ok, :float, context}
+  end
+
+  # "..."
+  def of_expr(literal, _stack, context) when is_binary(literal) do
+    {:ok, :binary, context}
+  end
+
+  # fn -> ... end
+  def of_expr(literal, _stack, context) when is_function(literal) do
+    {:ok, :dynamic, context}
+  end
+
+  # #PID<...>
+  def of_expr(literal, _stack, context) when is_pid(literal) do
```

Comment:
I think we can return `:pid` here but `reference` and `port` are not valid AST nodes, so you can remove the two clauses below. :)